File: OrchidCore/src/main/java/com/eden/orchid/api/theme/breadcrumbs/DefaultBreadcrumbStrategy.java
package com.eden.orchid.api.theme.breadcrumbs;

import com.eden.orchid.api.theme.pages.OrchidPage;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.stream.Collectors;

public final class DefaultBreadcrumbStrategy implements BreadcrumbStrategy {

    @Override
    public List<Breadcrumb> getBreadcrumbs(final OrchidPage page) {
        LinkedHashMap<OrchidPage, OrchidPage> breadcrumbPagesMap = new LinkedHashMap<>();

        OrchidPage parentPage = page;
        while(parentPage != null) {
            if(breadcrumbPagesMap.containsKey(parentPage)) break;

            breadcrumbPagesMap.put(parentPage, parentPage);
            parentPage = parentPage.getParent();
        }

        List<OrchidPage> breadcrumbPages = new ArrayList<>(breadcrumbPagesMap.values());
        Collections.reverse(breadcrumbPages);
        return breadcrumbPages
                .stream()
                .map(Breadcrumb::new)
                .collect(Collectors.toList())
                ;
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/themes/breadcrumbs/DefaultBreadcrumbStrategy.kt
package com.eden.orchid.impl.themes.breadcrumbs

import com.eden.orchid.api.theme.pages.OrchidPage
import java.util.ArrayList
import java.util.LinkedHashMap

class DefaultBreadcrumbStrategy : BreadcrumbStrategy {

    override fun getBreadcrumbs(page: OrchidPage): List<Breadcrumb> {
        val breadcrumbPagesMap = LinkedHashMap<OrchidPage, OrchidPage>()

        var parentPage: OrchidPage? = page
        while (parentPage != null) {
            if (breadcrumbPagesMap.containsKey(parentPage)) break // break out of a cycle

            breadcrumbPagesMap[parentPage] = parentPage
            parentPage = parentPage.parent
        }

        return ArrayList(breadcrumbPagesMap.values)
            .reversed()
            .map { Breadcrumb(it) }
    }
}
